keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document proposes to change the Prow presubmit handling to optionally version some or
+all presubmits in the same repository that also contains the code that is being tested.
+
+## Motivation
+
+Currently, all jobs for Prow are configured in a central `infra-config` repository that is
+in most cases distinct from the repositories whose code is being tested in presubmits. This
+poses severall challenges:
+
+* It is not possible to see in a Pull Request that introduces a new presubmit if that presubmit will
+  actually pass. The two workarounds for this are to either have both the job author and the reviewer
+  use two CLIs to create a Pod from the job or to make the Job initially optional, verify it with a
+  test Pull Request to the code repository, then make it mandatory. Both of these workarounds are
+  cumbersome.
+* The same issues apply when doing changes to the config of a presubmit
+* When a project maintains multiple branches, e.G. because there are release branches, the
+  maintainers must remember to create a copy of the presubmit in the `infra-config` repository.
+  Otherwise it is easely possible that the presubmit becomes incompatible with one of the branches
+  it is used in, because somone makes a change to its config and forgets to test against all branches.
+  This is an additional step maintainers have to remember when branching off.
+
+
+### Goals
```

Comment:
New blocking presubmits added via pull request trigger a run on all in-flight PRs 
Renamed blocking presubmits added via pull request trigger a migration on all in-flight PRs
Removed blocking presubmits via pull request trigger a status retire on all in-flight PRs